File: lib/song.dart
import 'dart:typed_data';

class Song {
  final String filePath;
  final String fileName;
  final String artist;
  final String title;

  Song(this.filePath, this.fileName, this.artist, this.title, this.minutes,
      this.seconds, this.image, this.album);

  final int minutes;
  final int seconds;
  final Uint8List? image;
  final String? album;

  factory Song.fromJson(Map json) {
    List<String> temp = (json['fileName'] as String).split("_");
    String title = json['title'] ??
        (json['fileName'] as String).substring(0, json['fileName'].length - 4);
    String artist = "Không rõ";
    if (temp.length == 3) {
      title = temp[0];
      artist = temp[1];
    }
    int duration = int.parse((json['duration'] as String));
    return Song(
      json['filePath'],
      json['fileName'],
      artist,
      title,
      duration ~/ 60000,
      (duration ~/ 1000) % 60,
      json['image'],
      json['album'],
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is Song && other.filePath == filePath;
  }

  @override
  int get hashCode {
    return filePath.hashCode ^
        fileName.hashCode ^
        artist.hashCode ^
        title.hashCode ^
        minutes.hashCode ^
        seconds.hashCode ^
        image.hashCode;
  }
}

File: android/app/src/main/kotlin/com/tankiem/flutter/flutter_mp3/Song.kt
package com.tankiem.flutter.flutter_mp3

import java.io.Serializable

data class Song(
    val filePath: String,
    val fileName: String,
    val image: ByteArray?,
    val artist: String?,
    val album: String?,
    val title: String?,
): Serializable {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Song

        if (filePath != other.filePath) return false
        if (fileName != other.fileName) return false
        if (image != null) {
            if (other.image == null) return false
            if (!image.contentEquals(other.image)) return false
        } else if (other.image != null) return false
        if (artist != other.artist) return false
        if (album != other.album) return false
        if (title != other.title) return false

        return true
    }

    override fun hashCode(): Int {
        var result = filePath.hashCode()
        result = 31 * result + fileName.hashCode()
        result = 31 * result + (image?.contentHashCode() ?: 0)
        result = 31 * result + (artist?.hashCode() ?: 0)
        result = 31 * result + (album?.hashCode() ?: 0)
        result = 31 * result + (title?.hashCode() ?: 0)
        return result
    }
}